### PR TITLE
kubeadm: refactor upgrade plan and apply commands

### PR DESCRIPTION
**Pull request by rojkov**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
The change

1. removes code duplication in `NewCmdApply()` and `NewCmdPlan()` by moving their common logic into `enforceRequirements()`;
2. makes `flags` provided by a user read-only structures and drops `SetImplicitFlags()`;
3. avoids state maintenance for non-interactiveness in flags.nonInteractiveMode, but rather calculates the mode from immutable variables.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Review by yagonobre — COMMENTED** [review]
Thanks @rojkov 
/priority important-longterm

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/upgrade/apply.go]: I think, that `cfg.KubernetesVersion` is going to have the old version here, since it's taken from the `ClusterConfiguration`, that is in a config map in the cluster.

Comment (rojkov) [reply cmd/kubeadm/app/cmd/upgrade/apply.go]: @rosti could you please elaborate a bit why this change is worse than the existing code?
In the old code `flags.newK8sVersionStr` is set one line above from the same source (`upgradeVars.cfg.KubernetesVersion`).

File: cmd/kubeadm/app/cmd/upgrade/common.go
Comment (rojkov) [reply cmd/kubeadm/app/cmd/upgrade/common.go]: Yes, makes sense. I've moved args handling in a separate function now.

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (rosti) [reply cmd/kubeadm/app/cmd/upgrade/apply.go]: Sorry about that, I must have been blind when I was looking at it. Totally missed the line above that.

**Comment by rosti** [comment]
@rojkov Can you squash most (if not all) of your commits here into a more general role?
This needs more reviews as it's size/L and I may have missed something.

/assign @fabriziopandini @neolit123 
/cc @yagonobre

**Review by neolit123 — APPROVED** [review]
thanks for the cleanup @rojkov 

unlike other projects, the k8s project usually doesn't follow the multi-commit per PR.
and we usually squash commits in a PR into one and focus a PR on a single change.
(i've seen exceptions).

with that in mind this PR probably could have been broken in 2-3 scoped single commit PRs.
having said that, i think it's OK to get this PR out of the way.
please, just keep that in mind for the future.
thanks.

LGTM
/lgtm
ping @fabriziopandini for the final `/approve`

**Comment by rojkov** [comment]
@rosti @neolit123 Thank you for review!
Indeed I was going to squash the commits after review, so I did it now and rebased.

Just in case the split commits can be found in https://github.com/rojkov/kubernetes/tree/kubeadm-refactor-enforceRequirements-split.

File: cmd/kubeadm/app/cmd/upgrade/common.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/upgrade/common.go]: Please, change `Infof` to `Infoln` here.

**Comment by rosti** [comment]
Actually, I spotted a few `Infof`s in `plan.go` and `apply.go`, that are not touched by this PR. They are out of the scope of it, but need to be fixed.
@rojkov given, that you work in this area, can we count on you for a followup PR fixing those? If you don't have the time I'll create a `help wanted` issue.

**Review by yagonobre — COMMENTED** [review]
Thanks @rojkov 
/lgtm

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (timothysc) [inline cmd/kubeadm/app/cmd/upgrade/apply.go]: I'd have to double check that the precedence order is correct, here.

**Comment by timothysc** [comment]
This should definitely be more than 1 PR per refactoring.  I'll defer to @fabriziopandini on approval.

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@rojkov thanks for this PR and sorry again for the delay on the review!

At first sight all the changes you are proposing seems reasonable, but I agree with @timothysc (and @neolit123) that probably it is better to break down all the different refactors in multiple PRs.

Please consider that ATM the CI signal for upgrades is not fully reliable, and the complexity of upgrades unfortunately already bite us in the past. Let's move carefully in this area (we are working on improving CI signal, but this will take some time).

PS. I see you are more and more active in kubeadm; I hope we will see you/hear your opinion at the kubeadm office hours soon!

**Comment by rojkov** [comment]
Submitted https://github.com/kubernetes/kubernetes/pull/73844 as a first PR in a series of split commits comprising this bigger PR.

**Comment by fabriziopandini** [comment]
/hold
as per comment above

**Comment by rojkov** [comment]
Closing as the content of this PR has been merged from separate PRs referenced above.

**Comment by neolit123** [comment]
thank you for the separate PRs @rojkov

**Comment by fabriziopandini** [comment]
Well done @rojkov

**Pull request closed without merging**
